File: src/main/java/de/bixilon/minosoft/data/text/TextComponent.java
/*
 * Minosoft
 * Copyright (C) 2020 Moritz Zwerger
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program.If not, see <https://www.gnu.org/licenses/>.
 *
 * This software is not affiliated with Mojang AB, the original developer of Minecraft.
 */

package de.bixilon.minosoft.data.text;

import de.bixilon.minosoft.Minosoft;
import de.bixilon.minosoft.config.ConfigurationPaths;
import de.bixilon.minosoft.protocol.protocol.ProtocolDefinition;
import de.bixilon.minosoft.util.Util;
import de.bixilon.minosoft.util.hash.BetterHashSet;
import javafx.animation.Animation;
import javafx.animation.KeyFrame;
import javafx.animation.Timeline;
import javafx.collections.ObservableList;
import javafx.scene.Node;
import javafx.scene.paint.Color;
import javafx.scene.text.Text;
import javafx.util.Duration;

import java.util.Objects;

public class TextComponent extends ChatComponent {
    private final String text;
    private RGBColor color;
    private BetterHashSet<ChatFormattingCode> formatting;

    public TextComponent(String text, RGBColor color, BetterHashSet<ChatFormattingCode> formatting) {
        this.text = text;
        this.color = color;
        this.formatting = formatting;
    }

    public TextComponent(String text, RGBColor color) {
        this.text = text;
        this.color = color;
    }

    public TextComponent(String text) {
        this.text = text;
    }

    public TextComponent setObfuscated(boolean obfuscated) {
        this.formatting.addOrRemove(PreChatFormattingCodes.OBFUSCATED, obfuscated);
        return this;
    }

    public TextComponent setBold(boolean bold) {
        this.formatting.addOrRemove(PreChatFormattingCodes.BOLD, bold);
        return this;
    }

    public TextComponent setStrikethrough(boolean strikethrough) {
        this.formatting.addOrRemove(PreChatFormattingCodes.STRIKETHROUGH, strikethrough);
        return this;
    }

    public TextComponent setUnderlined(boolean underlined) {
        this.formatting.addOrRemove(PreChatFormattingCodes.UNDERLINED, underlined);
        return this;
    }

    public TextComponent setItalic(boolean italic) {
        this.formatting.addOrRemove(PreChatFormattingCodes.ITALIC, italic);
        return this;
    }

    public TextComponent setReset(boolean reset) {
        this.formatting.addOrRemove(PostChatFormattingCodes.RESET, reset);
        return this;
    }

    public RGBColor getColor() {
        return this.color;
    }

    public TextComponent setColor(RGBColor color) {
        this.color = color;
        return this;
    }

    public BetterHashSet<ChatFormattingCode> getFormatting() {
        return this.formatting;
    }

    public TextComponent setFormatting(BetterHashSet<ChatFormattingCode> formatting) {
        this.formatting = formatting;
        return this;
    }

    @Override
    public int hashCode() {
        return Objects.hash(this.text, this.color, this.formatting);
    }

    @Override
    public boolean equals(Object obj) {
        if (super.equals(obj)) {
            return true;
        }
        if (hashCode() != obj.hashCode()) {
            return false;
        }
        TextComponent their = (TextComponent) obj;
        return this.text.equals(their.getMessage()) && this.color.equals(their.getColor()) && this.formatting.equals(their.getFormatting());
    }

    @Override
    public String toString() {
        return getANSIColoredMessage();
    }

    @Override
    public String getANSIColoredMessage() {
        StringBuilder builder = new StringBuilder();
        if (this.color != null) {
            builder.append(ChatColors.getANSIColorByRGBColor(this.color));
        }
        if (this.formatting != null) {
            this.formatting.forEach((chatFormattingCodes -> {
                if (chatFormattingCodes instanceof PreChatFormattingCodes code) {
                    builder.append(code.getANSI());
                }
            }));
        }
        builder.append(this.text);
        if (this.formatting != null) {
            this.formatting.forEach((chatFormattingCodes -> {
                if (chatFormattingCodes instanceof PostChatFormattingCodes code) {
                    builder.append(code.getANSI());
                }
            }));
        }
        builder.append(PostChatFormattingCodes.RESET);
        return builder.toString();
    }

    @Override
    public String getLegacyText() {
        StringBuilder output = new StringBuilder();
        Integer colorChar = ChatColors.getColorId(this.color);
        if (colorChar != null) {
            output.append(ProtocolDefinition.TEXT_COMPONENT_SPECIAL_PREFIX_CHAR).append(Integer.toHexString(colorChar));
        }
        this.formatting.forEach((chatFormattingCode -> output.append(ProtocolDefinition.TEXT_COMPONENT_SPECIAL_PREFIX_CHAR).append(chatFormattingCode.getChar())));
        output.append(this.text);
        output.append(ProtocolDefinition.TEXT_COMPONENT_SPECIAL_PREFIX_CHAR).append(PostChatFormattingCodes.RESET.getChar());
        return output.toString();
    }

    @Override
    public String getMessage() {
        return this.text;
    }

    @Override
    public ObservableList<Node> getJavaFXText(ObservableList<Node> nodes) {
        Text text = new Text(this.text);
        text.setFill(Color.WHITE);
        if (Minosoft.getConfig().getBoolean(ConfigurationPaths.BooleanPaths.CHAT_COLORED) && this.color != null) {
            text.setFill(Color.web(this.color.toString()));
        }
        this.formatting.forEach((chatFormattingCode -> {
            if (chatFormattingCode instanceof PreChatFormattingCodes code) {
                switch (code) {
                    case OBFUSCATED -> {
                        // ToDo: potential memory leak: Stop timeline, when TextComponent isn't shown anymore
                        Timeline obfuscatedTimeline;
                        if (Minosoft.getConfig().getBoolean(ConfigurationPaths.BooleanPaths.CHAT_OBFUSCATED)) {
                            obfuscatedTimeline = new Timeline(new KeyFrame(Duration.millis(50), e -> {
                                char[] chars = text.getText().toCharArray();
                                for (int i = 0; i < chars.length; i++) {
                                    chars[i] = Util.getRandomChar(ProtocolDefinition.OBFUSCATED_CHARS);
                                }
                                text.setText(new String(chars));
                            }));
                        } else {
                            obfuscatedTimeline = new Timeline(new KeyFrame(Duration.millis(500), e -> text.setVisible(false)), new KeyFrame(Duration.millis(1000), e -> text.setVisible(true)));
                        }
                        obfuscatedTimeline.setCycleCount(Animation.INDEFINITE);
                        obfuscatedTimeline.play();
                        text.getStyleClass().add("obfuscated");
                    }
                    case BOLD -> text.setStyle("-fx-font-weight: bold;");
                    case STRIKETHROUGH -> text.setStyle("-fx-strikethrough: true;");
                    case UNDERLINED -> text.setStyle("-fx-underline: true;");
                    case ITALIC -> text.setStyle("-fx-font-weight: italic;");
                }
            }
        }));
        nodes.add(text);
        return nodes;
    }
}

File: src/main/java/de/bixilon/minosoft/data/text/TextComponent.kt
/*
 * Minosoft
 * Copyright (C) 2020-2025 Moritz Zwerger
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU General Public License as published by the Free Software Foundation, either version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with this program. If not, see <https://www.gnu.org/licenses/>.
 *
 * This software is not affiliated with Mojang AB, the original developer of Minecraft.
 */
package de.bixilon.minosoft.data.text

import de.bixilon.kutil.enums.BitEnumSet
import de.bixilon.kutil.json.MutableJsonObject
import de.bixilon.minosoft.data.registries.identified.ResourceLocation
import de.bixilon.minosoft.data.text.events.click.ClickEvent
import de.bixilon.minosoft.data.text.events.hover.HoverEvent
import de.bixilon.minosoft.data.text.formatting.FormattingCodes
import de.bixilon.minosoft.data.text.formatting.TextStyle
import de.bixilon.minosoft.data.text.formatting.color.ChatColors
import de.bixilon.minosoft.data.text.formatting.color.RGBAColor
import de.bixilon.minosoft.data.text.formatting.color.RGBColor
import de.bixilon.minosoft.protocol.protocol.ProtocolDefinition


open class TextComponent(
    message: Any? = "",
    override var color: RGBAColor? = null,
    override val formatting: BitEnumSet<FormattingCodes> = FormattingCodes.set(),
    var font: ResourceLocation? = null,
    var clickEvent: ClickEvent? = null,
    var hoverEvent: HoverEvent? = null,
) : ChatComponent, TextStyle {
    override var message: String = message.toString().replace(ProtocolDefinition.TEXT_COMPONENT_FORMATTING_PREFIX, '&')

    override fun obfuscate(): TextComponent {
        formatting += FormattingCodes.OBFUSCATED; return this
    }

    override fun bold(): TextComponent {
        formatting += FormattingCodes.BOLD; return this
    }

    override fun strikethrough(): TextComponent {
        formatting += FormattingCodes.STRIKETHROUGH; return this
    }

    override fun underline(): TextComponent {
        formatting += FormattingCodes.UNDERLINED; return this
    }

    override fun italic(): TextComponent {
        formatting += FormattingCodes.ITALIC; return this
    }

    fun clickEvent(clickEvent: ClickEvent?): TextComponent {
        this.clickEvent = clickEvent
        return this
    }

    fun hoverEvent(hoverEvent: HoverEvent?): TextComponent {
        this.hoverEvent = hoverEvent
        return this
    }

    fun color(color: RGBColor) = this.color(color.rgba())
    fun color(color: RGBAColor): TextComponent {
        this.color = color
        return this
    }

    override fun toString(): String {
        return legacy
    }

    override fun setFallbackColor(color: RGBAColor): TextComponent {
        if (this.color == null) {
            this.color = color
        }
        return this
    }

    override val ansi: String
        get() {
            val builder = StringBuilder()
            this.color?.let {
                builder.append(it.ansi)
            }

            for (formattingCode in this.formatting) {
                builder.append(formattingCode.ansi)
            }
            builder.append(this.message)
            builder.append(FormattingCodes.RESET.ansi)
            return builder.toString()
        }

    override val legacy: String
        get() {
            val builder = StringBuilder()
            ChatColors.getChar(color)?.let { builder.append(ProtocolDefinition.TEXT_COMPONENT_FORMATTING_PREFIX).append(it) }
            for (formattingCode in this.formatting) {
                builder.append(ProtocolDefinition.TEXT_COMPONENT_FORMATTING_PREFIX)
                builder.append(formattingCode.char)
            }
            builder.append(this.message)
            builder.append(ProtocolDefinition.TEXT_COMPONENT_FORMATTING_PREFIX).append(FormattingCodes.RESET.char) // ToDo: This should not always be appended
            return builder.toString()
        }

    override fun toJson(): Any {
        if (message.isEmpty()) {
            return emptyMap<String, Any>()
        }
        val json: MutableJsonObject = mutableMapOf(
            "text" to message
        )

        for (formatting in formatting) {
            json[formatting.json] = true
        }

        color?.let { json["color"] = ChatColors.NAME_MAP.getKey(it) ?: it.toString() }

        // TODO: hover, click event

        return json
    }

    fun copy(message: Any? = this.message, color: RGBAColor? = this.color, formatting: BitEnumSet<FormattingCodes> = this.formatting, clickEvent: ClickEvent? = this.clickEvent, hoverEvent: HoverEvent? = this.hoverEvent) = TextComponent(
        message = message,
        color = color,
        formatting = formatting,
        clickEvent = clickEvent,
        hoverEvent = hoverEvent,
    )

    override val length: Int
        get() = message.length

    override fun getTextAt(pointer: Int): TextComponent {
        if (pointer < 0 || pointer > message.length) {
            throw IllegalArgumentException("Pointer out of bounds: $pointer")
        }
        return this
    }

    override fun cut(length: Int) {
        if (length <= 0) {
            throw IllegalArgumentException("Can not cut <= 0: $length")
        }
        if (length >= message.length) {
            throw IllegalArgumentException("Can not cut beyond length: $length >= ${message.length}")
        }

        message = message.substring(0, length)
    }

    override fun copy(): ChatComponent {
        return TextComponent(message, color, formatting.copy(), font, clickEvent, hoverEvent)
    }

    override fun hashCode(): Int {
        return message.hashCode()
    }

    override fun equals(other: Any?): Boolean {
        if (other === this) {
            return true
        }
        if (other !is TextComponent) {
            return false
        }
        return message == other.message && color == other.color && formatting == other.formatting && clickEvent == other.clickEvent && hoverEvent == other.hoverEvent
    }

    override fun trim(): ChatComponent? {
        if (message.isEmpty()) return null
        return this
    }
}
